homeassistant/components/homekit/accessories.py

```diff
@@ -104,24 +111,42 @@ def __init__(self, hass, driver, name, entity_id, aid, config,
         async_track_state_change(
             self.hass, self.entity_id, self.update_state_callback)
 
+        if self.linked_battery_sensor:
+            self.hass.async_add_job(self.update_linked_battery, None, None,
+                                    state)
+            async_track_state_change(
+                self.hass, self.linked_battery_sensor,
+                self.update_linked_battery)
+
     @ha_callback
     def update_state_callback(self, entity_id=None, old_state=None,
                               new_state=None):
         """Handle state change listener callback."""
         _LOGGER.debug('New_state: %s', new_state)
         if new_state is None:
             return
-        if self._support_battery_level:
+        if self._support_battery_level and not self.linked_battery_sensor:
             self.hass.async_add_executor_job(self.update_battery, new_state)
         self.hass.async_add_executor_job(self.update_state, new_state)
 
+    @ha_callback
+    def update_linked_battery(self, entity_id=None, old_state=None,
+                              new_state=None):
+        """Handle linked battery sensor state change listener callback."""
+        self.update_battery(new_state)
```

Comment:
Since `update_battery` is a sync method, it needs to be executed with `async_add_executer_job`.
See `update_state_callback`.